### PR TITLE
WebSockets

### LINKED ISSUE #182: WSS integration
Hello, i have a wss server that i want to connect to. Does this library support wss connections?

### PR DESCRIPTION
**Pull request by dnKaratzas**
- Add Network Provider
- Add HttpNetworkProvider
- Add WebSocketNetworkProvider
- Remove Deprecated Methods

EthereumClients:
- Add EthereumHttpClient
- Add EthereumWebsocketClient

Resolves  #182, #240, #241

TODO:
- Update Readme

### CODE REVIEW

File: web3swift/src/Client/EthereumClient.swift [hunk 1]
```diff
@@ -8,100 +8,20 @@
...
-    case tooManyResults
-    case executionError(JSONRPCErrorDetail)
-    case unexpectedReturnValue
-    case noResultFound
-    case decodeIssue
-    case encodeIssue
-    case noInputData
-}
-
 public class EthereumClient: EthereumClientProtocol {
```
Comment (dnKaratzas) [inline web3swift/src/Client/EthereumClient.swift:17]: @DarthMike Should I rename that to `AbstractEthereumClient`? The initializer is internal but should we give a more meaningful name?

Comment (DarthMike) [reply web3swift/src/Client/EthereumClient.swift:17]: `BaseEthereumClient` ?

File: web3sTests/ENS/ENSTests.swift [hunk 2]
```diff
@@ -30,7 +30,7 @@ class ENSTests: XCTestCase {
 
             let tx = try function.transaction()
 
-            let dataStr = try await client?.eth_call(tx, block: .Latest)
+            let dataStr = try await client?.eth_call(tx, resolution: .noOffchain(failOnExecutionError: true), block: .Latest)
```
[the remark this answers is not in the document]
Comment (dnKaratzas) [reply web3sTests/ENS/ENSTests.swift:33]: I must cast the Protocol to the actual type cause the protocol don't have defaults

File: web3swift/src/Client/EthereumClient.swift (hunk 1 above)
Comment (dnKaratzas) [reply web3swift/src/Client/EthereumClient.swift:17]: Sounds better thanks

File: web3sTests/OffchainLookup/OffchainLookupTests.swift [hunk 3]
```diff
@@ -160,7 +160,7 @@ class OffchainLookupTests: XCTestCase {
         let tx = try! function.transaction()
 
         do {
-            let _ = try await client.eth_call(tx)
+            let _ = try await client.eth_call(tx, resolution: .noOffchain(failOnExecutionError: true), block: .Latest)
```
[the remark this answers is not in the document]
Comment (dnKaratzas) [reply web3sTests/OffchainLookup/OffchainLookupTests.swift:163]: Let's keep it like that cause the logger have a function called log

File: web3swift/src/Account/EthereumAccount.swift [hunk 4]
```diff
@@ -29,6 +30,7 @@ public enum EthereumAccountError: Error {
 public class EthereumAccount: EthereumAccountProtocol {
     private let privateKeyData: Data
     private let publicKeyData: Data
+    private let logger: Logger
```
[the remark this answers is not in the document]
Comment (dnKaratzas) [reply web3swift/src/Account/EthereumAccount.swift:33]: Let's keep it like that cause the logger have a function called log

File: web3sTests/ENS/ENSTests.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (DarthMike) [reply web3sTests/ENS/ENSTests.swift:33]: 👍

**Review by DarthMike — APPROVED** [review]

**Comment by DarthMike** [comment]
Good work @dnKaratzas !

**Pull request merged**